### ISSUE TITLE
Breakdown of this proposal

### ISSUE THREAD

**Issue by littledan** [erights/PLNSOWNSF#1]
I see a few things being proposed here. While I like the big-picture exploration of how things look when all put together, I want to break down the individual components here to help us figure out next steps. Though I see you've illustrated a way that these could work together, I see them as somewhat orthogonal

## External private names in field declarations

This is a real thing that's being provided that the class fields proposal currently lacks. We've discussed a number of different syntaxes for this; we've also discussed whether the reference should come from the private identifier scope or from the lexical scope. Seems like this proposal chooses lexical, privileging generality over simplicity for limited cases, which I am sympathetic with.

I think we should strongly consider this feature as a follow-on to the private fields and methods proposal.

## `PrivateName.prototype.init` method

As we've discussed, the omission of this method was a deliberate design decision: the goal is to encourage programmers to use field declarations, which encourage more static shape, rather than encouraging private fields to be added to objects later on. Decorators give a metaprogramming interface to add additional private fields to classes, rather than instances. Do you think this might meet some of the design goals of this proposal?

## `::` syntax

I have to say, I'm just not a fan of this change. On a first glance, it looks prettier to avoid the `#` character, but to me, `::` reads more as an infix operator than as part of the name. I think it's important to the hazard one infix operator is substituted for another (which would end up leaking your private state!). Thinking of `#` as part of the name rather than the operator is our mnemonic that makes this easy. @ljharb has expressed similar sentiments.

## Private name protocol providing hook-ability

In terms of expressiveness, this seems similar to the ability of decorators to turn a private field into a private accessor pair, modulo the ability to share private names for several uses. Or is there more to it than this?

## Private in object literals

I like this part of the proposal, and I think it layers well as a follow-on to the private fields proposal. In particular, you seem to have made the design decision that, while `class A { ::b = c }` creates a new private name binding, object literals *don't* create a new binding with `{ ::b : c }`. After initially taking the opposite position and a bit of thought, I share this intuition.

**Comment by bathos-wistia** [comment]
> As we've discussed, the omission of this method was a deliberate design decision: the goal is to encourage programmers to use field declarations, which encourage more static shape, rather than encouraging private fields to be added to objects later on. Decorators give a metaprogramming interface to add additional private fields to classes, rather than instances. Do you think this might meet some of the design goals of this proposal?

The desire to encourage a particular pattern makes sense to me. However “encourage” seems to be a euphemism here — forbidding alternatives to foo is not typically called “encouraging” foo :)

Is class-based private field initialization not still being encouraged through its simpler syntax and API?

**Comment by littledan** [comment]
There's no strong prohibition, as you can use the super return trick to add private fields to existing objects. That's why it's just encouragement.

**Comment by bathos-wistia** [comment]
The super return trick does require declaration of the fields in a class — just a different one than the one(s) that will obtain the fields. Unless I’m misunderstanding, reified private names could not be used for this if init were omitted.

A bit unintuitively, I think that means that, yes, there would still be a way to add fields to arbitrary objects at times other than their initialization, but it would be available only by (ab)using the API which is intended for class usage, instead of via the more primitive API. That seems weird, no?

Worth noting also that init facilitates setting fields on Proxy objects without using the super return trick.

**Comment by littledan** [comment]
No disagreement with your analysis, but I'm not convinced that those facts about what can be expressed mean we should provide an `init` method.

**Comment by hax** [comment]
> :: syntax ... I think it's important to the hazard one infix operator is substituted for another (which would end up leaking your private state!). Thinking of # as part of the name rather than the operator is our mnemonic that makes this easy.

@littledan Yes, substitution `.` to `::` may be the "hazard", but `obj.#foo` is also have hazards:

-  "Thinking of # as part of the name rather than the operator" cause programmers expect `obj.#foo` still follow property lookup semantic, but actually not.
- Disgust of `#` notation which may cause community break

Which "hazard" is more "hazard"? It may be subjective. 

I just want to repeat my position again, the worst hazard is syntax -> semantic ambiguity

- `[x] = [1,2,3]` has totally different semantic in class body
- `a = 1; b = a + 1` has totally different semantic in class body
- `foo.#bar` and `foo.bar`(also `foo._bar`)  has totally different semantic

One proposal have three such worst hazards, enough for me to veto it.
